site/content/blog/2020-12-04-common-i18n-mistakes.md

```diff
@@ -0,0 +1,302 @@
+
+When you reply to an older message in Mattermost, we display "Commented on Billy's message:" to tell the user what you're responding to, but for a while after we had translation support added, the code for that looked like this:
+
+```jsx
+let username = '@' + rootPost.username;
+if (!username.endsWith('s')) {
+    username += '\'s';
+}
+
+return (
+    <FormattedMessage
+        id='post.commentedOn'
+        defaultMessage='Commented on {username} message:'
+        values={{username}}
+    />
+)
+```
+
+Here, we're using react-intl to add the user's username into the message as a value, but we're making a mistake here by trying to make the username into a possessive adjective. By manually adding the "'s", we're applying English grammar regardless of the user's language setting, so in other languages, we're adding text that doesn't make sense.
+
+We solved this by moving the "'s" into the translation string. That required changing our grammar rules to add "'s" to the end of names that end in the letter S, but that's still correct by some grammar guidelines. Isn't English great?
+
+```jsx
+return (
+    <FormattedMessage
+        id='post.commentedOn'
+        defaultMessage="Commented on @{username}'s message:"
+        values={{username: rootPost.username}}
+    />
+);
+```
+
+## Mistake 3: Incorrect pluralization
+
+Another common mistake is to forget that pluralization exists or to attempt to avoid it by adding (s) to the end of a word. While the latter is technically correct, it doesn't look very nice, and thankfully, our translation libraries can help us out there.
+
+Say you want to write something that sends an email to a list of any number of people, and you want it to report back on its progress as it sends those email in both the UI and in the server's logs.
+
+```jsx
+return (
+    <FormattedMessage
+        id='email_sender.remaining'
+        defaultMessage='{remaining, number} email(s) remaining.'
+        values={{remaining}}
+    />
+);
+```
+
+```go
+/*
+    The translation file contains:
+    {
+        "id": "app.email_sender.remaining",
+        "translation": "{{.Remaining}} email(s) remaining"
+    }
+*/
+
+func logEmailsRemaining(remaining int) {
+    log.Print(translateFunc("app.email_sender.remaining", map[string]interface{}{
+        "Remaining": remaining,
+    }))
+}
+```
+
+Note that the React example specifies that `remaining` is a number since otherwise zero would be treated as an empty string because it's falsey.
+
+As mentioned before, these are both technically correct since they add the number of remaining emails to the right place, but they both look ugly. Instead, both libraries support pluralization by passing in a count of some object which is the number of remaining emails in this case.
+
+```jsx
+return (
+    <FormattedMessage
+        id='email_sender.remaining'
+        defaultMessage='{remaining, number} {remaining, plural, one {email} other {emails}} remaining.'
```

Comment:
not sure if you want to add this as an example in Spanish there is a plural word for `remaining`.
singular: restante
plural: restantes

Although you can translate it to something like
`{remaining, number} {remaining, plural, one {correo pendiente} other {correos pendientes}}.`

Reply:
Good idea. That'll also help reinforce how other languages do things differently. I considered trying to add some non-English examples myself, but my French is really rusty.